infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java

```diff
@@ -434,6 +452,21 @@ public void onNext(CorfuStreamEntries results) {
                 log.info("onNext :: operation={}, key={}, payload={}, metadata={}", entry.getOperation().name(),
                         entry.getKey(), entry.getPayload(), entry.getMetadata());
             }
+
+            // If new topologyConfig is applied, block until it takes effect (topologyConfig in DefaultClusterManager is updated)
+            if (clusterManager.configId != lastConfigId) {
+                try {
+                    clusterManager.topologyUpdateLock.lock();
+                    while (clusterManager.topologyConfig.getTopologyConfigID() != lastConfigId + 1) {
```

Comment:
Just confirming if this loop for handling spurious wake in condition variable?

Reply:
In my opinion the atomicity we want to achieve here is: when the stream listener consumes a topology change msg, it guarantees the topology in CorfuReplicationClusterManagerBaseAdapter is actually updated (before exit the onNext() method), but it won't block until the consumption of this event in discovery service, which I think also aligns with the real use case?

From the correctness perspective, I think we are fine here. As the topology in the adapter in the source of truth for new topology, and discovery service is doing some role-based setup according to it. Also, as mentioned in previous comment, our implementation of this DefaultClusterManager also fetches topology from the adapter.